feat: Language translations finder and update

neznaika0 opened the pull request:
**Description**
See thread https://forum.codeigniter.com/showthread.php?tid=88299

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

neznaika0 commented:
I fixed typehint for PHP 7.4

kenjis commented:
Fix broken tests. https://github.com/codeigniter4/CodeIgniter4/actions/runs/6057011818/job/16438443140?pr=7889

kenjis commented on tests/system/Commands/Translation/LocalizationFinderTest.php:
```diff
@@ -0,0 +1,201 @@
+use Locale;
+
+/**
+ * @internal
```
Add `@group Others`.

```
1) CodeIgniter\AutoReview\FrameworkCodeTest::testEachTestClassHasCorrectGroupAnnotation with data set "CodeIgniter\Commands\Translation\LocalizationFinderTest" ('CodeIgniter\Commands\Translat...erTest')
[CodeIgniter\Commands\Translation\LocalizationFinderTest] Test class is missing a @group annotation.
```
https://github.com/codeigniter4/CodeIgniter4/actions/runs/6057011814/job/16438441858?pr=7889

kenjis commented:
Run `vendor/bin/rector {file_path}`.
https://github.com/codeigniter4/CodeIgniter4/actions/runs/6057011803/job/16438441396?pr=7889

neznaika0 commented:
There is a conflict between the Rector and php-cs-fixer (pre-commit). The rector erases the indents in the method, and php-cs-fixer adds.
```
      ---------- begin diff ----------
--- /home/aleksandr/www/codeigniter/tests/system/Commands/Translation/LocalizationFinderTest.php
+++ /home/aleksandr/www/codeigniter/tests/system/Commands/Translation/LocalizationFinderTest.php
@@ -149,33 +149,33 @@
     private function getActualTableWithNewKeys(): string
     {
         return <<<'TEXT_WRAP'
-+------------------+----------------------------------------------------+
-| File             | Key                                                |
-+------------------+----------------------------------------------------+
-| TranslationOne   | TranslationOne.Copyright                           |
...many lines
-| TranslationThree | TranslationThree.formFields.new.short_tag          |
-+------------------+----------------------------------------------------+
-TEXT_WRAP;
+            +------------------+----------------------------------------------------+
+            | File             | Key                                                |
+            +------------------+----------------------------------------------------+
+            | TranslationOne   | TranslationOne.Copyright                           |
...many lines
+            | TranslationThree | TranslationThree.formFields.new.short_tag          |
+            +------------------+----------------------------------------------------+
+            TEXT_WRAP;
     }
 
     private function realizeAssertion(): void

      ----------- end diff -----------
```

kenjis commented:
The conflicts are not a problem. We always need to run `composer cs-fix` after refactoring by rector.

kenjis commented:
> Should I move files from Controllers/Translation to another location? This violates the router's tests

If possible, it is better.
Or update the failed tests.

neznaika0 commented:
I moved it to another folder with examples.
Added some description in the guide.

Is it possible to get an assessment of the work, is it worth continuing? Or will the function be rejected after the tests?

kenjis commented:
We have not yet reviewed the contents of the code because the GitHub Action checks have not passed.

However, this feature is useful and I would like to include it in 4.5.

ddevsr commented on system/Commands/Translation/LocalizationFinder.php:
```diff
@@ -0,0 +1,339 @@
+            $countNewKeys += $this->arrayCountRecursive($languageDiff);
+
+            if ($cliOptionShowNew) {
+                $tableRows = array_merge($this->arrayToTableRows($langFileName, $languageDiff), $tableRows);
```
I would lke use

```suggestion
                $tableRows = [$this->arrayToTableRows($langFileName, $languageDiff), $tableRows];
```

Ref : https://github.com/rectorphp/rector/blob/main/docs/rector_rules_overview.md#arraymergeofnonarraystosimplearrayrector

neznaika0 replied:
No. Read the instructions carefully. 
You are trying to merge two arrays, and the documentation talks about strings. This will result in an error.
+ I applied the Rector before the commit

neznaika0 commented:
Oops. My commit has merged with the develop branch. Is it ugly?
I added to PR dashed keys search as `lang('Lang-File.dashed.key-with-dash')`
